api/v1_users_balance_history.go

```diff
@@ -0,0 +1,119 @@
+package api
+
+import (
+	"time"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/jackc/pgx/v5"
+)
+
+type BalanceHistoryQueryParams struct {
+	StartTime *time.Time `query:"start_time"`
+	EndTime   *time.Time `query:"end_time"`
+}
+
+type BalanceHistoryDataPoint struct {
+	Timestamp  int64   `json:"timestamp"`
+	BalanceUsd float64 `json:"balance_usd"`
+}
+
+type BalanceHistoryResponse struct {
+	Data []BalanceHistoryDataPoint `json:"data"`
+}
+
+// v1UsersBalanceHistory returns the historical balance data for a user
+// Returns hourly data for ≤7 days, daily data for >7 days
+func (app *ApiServer) v1UsersBalanceHistory(c *fiber.Ctx) error {
+	// Get user ID from middleware (set by requireUserIdMiddleware)
+	userId := app.getUserId(c)
+
+	// Parse query params
+	params := BalanceHistoryQueryParams{}
+	if err := c.QueryParser(&params); err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid query parameters")
+	}
+
+	// Set default time range if not provided (last 7 days)
+	now := time.Now()
+	endTime := now
+	startTime := now.Add(-7 * 24 * time.Hour)
+
+	if params.StartTime != nil {
+		startTime = *params.StartTime
+	}
+	if params.EndTime != nil {
+		endTime = *params.EndTime
+	}
+
+	// Validate time range
+	if endTime.Before(startTime) {
+		return fiber.NewError(fiber.StatusBadRequest, "end_time must be after start_time")
+	}
+
+	// Determine granularity based on time range
```

Comment:
Super nit but maybe we have a separate query param for max flexibility? And maaaybe if we think it's too expensive we could validate that you don't do sub-hourly for > 7 days?